src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs

```diff
@@ -141,13 +141,66 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
                         throw;
                     }
                 }
-                await Task.Delay(20, cancellationToken).ConfigureAwait(false);
+                await Task.Delay(MillisecondsInterval, cancellationToken).ConfigureAwait(false);
             }
             throw new InvalidOperationException();
         }
 
-        private void FileChanged(object sender, FileSystemEventArgs e)
+        private async void FileChanged(object sender, FileSystemEventArgs e)
```

Comment:
I'd probably add some measure to have only single in-flight async executing handler at a time (e.g. slim MRE or Mutex, attempted to be waited with 0 interval and just return if not acquired).

IIRC the updating code might be updating multiple packages while holding the lock.
This way we might get quite some amount of in flight tasks tryig to acquire the lock and send the notification, while the subscriber of the even will as well try to acquire the lock. There will be a heavy Tasks and synchronization utilization, while actually only the last invocation matters.

Reply:
Unfortunately in FileChanged event calling LockAsync() failed to get the lock with `System.InvalidOperationException: Lock is already taken.` in CI run after retries. It might be caused by heavy tasks trying to acquire the lock as you mentioned. 

In addition, removing `SettingsChanged?.Invoke()` from `SetInstalledTemplatePackagesAsync` caused other problem. Many Microsoft.TemplateEngine.IDE.IntegrationTests such as Create_BasicTest_Package failed. I need to figure it out. Probably though `SettingsChanged?.Invoke()` is executed in FileChanged() event anyway, there is some delay especially with retries acquiring the lock or checking the state of lock. The delay caused something wrong.

Reply:
Resolved now